File: sujet4_bang_bang/lib/sfun.m
function s  = sfun(y,options,par)
%-------------------------------------------------------------------------------------------
%
%    sfun
%
%    Description
%
%        Computes the shooting function
%
%-------------------------------------------------------------------------------------------
%
%    Matlab
%
%        s = sfun(y, options, par)
%
%    Inputs
%
%        y       - real vector  : shooting variable
%        options - struct       : odeset options
%        par     - real vector  : parameters, par=[] if no parameters
%
%    Outputs
%
%        s       - real vector, shooting value
%
%-------------------------------------------------------------------------------------------
%
% par = [t0; tf; x0; xf]
% y   = [p0; t1; z1]
%
t0 = par(1);
tf = par(2);
x0 = par(3);
xf = par(4);

n  = 1; %dimension of te state
p0 = y(1:n);
z0 = [x0;p0];

t1 = y(n+1);
z1 = y(n+2:n+2+2*n-1);

%% A REMPLACER
[ ~, Za ]  = exphvfun([t0 t1], z0, options, par, 1);
[ ~, Zb ]  = exphvfun([t1 tf], z1, options, par, 2);
s = [ xf - Zb(1:n,end); norm(z1(n+1:end,end))-1;z1 - Za(:,end)];

File: sujet5_hampath/Probleme_P4epsilon/sfun.f90
!!********************************************************************
!>  \ingroup problemDefinition
!!  \brief Shooting function.
!!  \param[in] ny       Shooting variable dimension
!!  \param[in] y        Shooting variable
!!  \param[in] npar     Number of optional parameters
!!  \param[in] par      Optional parameters
!!
!!  \param[out] s       Shooting value, same dimension as y
!!  \attention          The vector par can be used for constant values
!!                      or homotopic parameters.
!!
!!  \author Olivier Cots (INP-ENSEEIHT-IRIT)
!!  \date   2015-2016
!!  \copyright Eclipse Public License
    Subroutine sfun(ny,y,npar,par,s)
        use mod_exphv4sfun
        implicit none
        integer,                            intent(in)  :: ny
        integer,                            intent(in)  :: npar
        double precision,  dimension(ny),   intent(in)  :: y
        double precision,  dimension(npar), intent(in)  :: par
        double precision,  dimension(ny),   intent(out) :: s

        external printandstop

        !!  The user can call hfun and exphv subroutines inside sfun
        !!
        !!  hfun: code the true Hamiltonian
        !!
        !!      syntax: call hfun(t,n,z,iarc,npar,par,h)
        !!
        !!      see hfun.f90 for details.
        !!
        !!  exphv: computes the chronological exponential of the
        !!          Hamiltonian vector field hv defined by h.
        !!
        !!      syntax: call exphv(tspan,n,z0,iarc,npar,par,zf)
        !!
        !!       integer         , intent(in)    :: n
        !!       integer         , intent(in)    :: iarc
        !!       integer         , intent(in)    :: npar
        !!       double precision, intent(in)    :: tspan(:) = [t0 ... tf]
        !!       double precision, intent(in)    :: z0(2*n)
        !!       double precision, intent(in)    :: par(npar)
        !!       double precision, intent(out)   :: zf(2*n) = z(tf)
        !!

        !local variables
        double precision :: z0(4), zf(4), tspan(2)
        integer          :: n, iarc

        IF (ny.NE.1) THEN
            call printandstop('Error: wrong shooting variable dimension.')
        END IF
        IF (npar.NE.5) THEN
            call printandstop('Error: wrong par dimension.')
        END IF

        n        = 1                ! Dimension of the state

        tspan(1) = par(1)           ! t0
        tspan(2) = par(2)           ! tf

        z0(  1)  = par(3)           ! x0
        z0(  2)  = y(1)                       

        iarc     = 1                

        call exphv(tspan,n,z0,iarc,npar,par,zf)

        s(1)     = zf(1) - par(4)   ! x(tf) - xf


    end subroutine sfun